server/src/Cache.ts

```diff
@@ -0,0 +1,135 @@
+import fs from 'fs'
+import path from 'path'
+import serverconfig from './serverconfig.js'
+
+type CacheOpts = {
+	/** Set a specific directory within the cache (e.g. bam, gsea, etc.) instead of
+	 * the default/root cachedir.*/
+	cachedir?: string
+	/** time to wait before triggering another call to mayDeleteCacheFiles(),
+	 * this is used to debounce/prevent multiple active calls to mayDeleteCacheFiles()
+	 * also assumed to be roughly equivalent to the minimum required time for a file read
+	 * to complete, otherwise deleting sooner than this may cause a file read error;
+	 * this last assumption only applies to file deletion when the maxSize is exceeded */
+	checkWait?: number
+	/** file extensions to target. Leave blank to delete everything in the dir */
+	fileExtensions?: string[]
+	/** the max age for the modified time, will delete files whose modified time
+	 * exceeds this "aged" access (in milliseconds) */
+	maxAge?: number
+	/** maximum allowed cache size in bytes */
+	maxSize?: number
+}
+
+/**
+ * This is a prototype cache class for managing files in the cachedir.
+ */
+
+export class Cache {
+	/** a pending timeout reference from setTimeout that calls mayDeleteCacheFiles */
+	#cacheCheckTimeout: NodeJS.Timeout | undefined | number
+	cachedir: string
+	public checkWait: number
+	#fileExtensions: string[]
+	/** in milliseconds */
+	#maxAge: number
+	/** in bytes */
+	#maxSize: number
+	#nextCheckTime = 0
+
+	constructor(opts: CacheOpts = {}) {
+		this.#cacheCheckTimeout = 0
+		this.cachedir = opts.cachedir || serverconfig.cachedir
+		this.checkWait = opts.checkWait || 1 * 60 * 1000
+		this.#fileExtensions = opts.fileExtensions || []
+		this.#maxAge = opts.maxAge || 2 * 60 * 60 * 1000
+		this.#maxSize = opts.maxSize || 5e9
+	}
+
+	mayResetCacheCheckTimeout(wait = 0) {
+		// do not trigger the cache check when only validating the server
+		if (process.argv.includes('validate')) return
+
+		const checkTime = Date.now() + wait
+		if (this.#cacheCheckTimeout) {
+			if (this.#nextCheckTime && this.#nextCheckTime <= checkTime + 5) return
+			else {
+				clearTimeout(this.#cacheCheckTimeout)
+				this.#cacheCheckTimeout = undefined
+			}
+		}
+		this.#nextCheckTime = checkTime
+		console.log(`will trigger mayDeleteCacheFiles() in ${wait} ms`)
+		this.#cacheCheckTimeout = setTimeout(this.mayDeleteCacheFiles.bind(this), wait)
+	}
+
+	async mayDeleteCacheFiles() {
+		console.log(`checking for cached files to delete ...`)
+		try {
+			const minTime = Date.now() - this.#maxAge
+			const filenames = await fs.promises.readdir(this.cachedir)
+			const files: { path: any; time: any; size: any; deleted?: any }[] = [] // keep list of undeleted files. may need to rank them and delete old ones ranked by age
+			let totalSize = 0,
+				deletedSize = 0,
+				totalCount = 0,
+				deletedCount = 0
+			for (const filename of filenames) {
+				if (this.#fileExtensions.length && !this.#fileExtensions.some(ext => filename.endsWith(ext))) continue
```

Comment:
Should simplify with `this.fileExtensions.includes(path.extname(filename))`, instead of creating and running a callback for each array entry for each loop iteration.  Note that the entries would need to include a dot (`.txt`) when using `path.extname`.  May also consider setting `fileExtensions` as a `Set`, so that it's guaranteed all fileExtension entries are unique and the check becomes `this.fileExtensions.has(path.extname(filename))`.